MNT Add code scanning workflow

lesteve opened the pull request:
Folllowing https://github.com/scikit-learn/scikit-learn/issues/28151, I started to read a few security-related blog posts, for example https://securitylab.github.com/research/github-actions-untrusted-input/.

This PR would add [Code scanning](https://docs.github.com/en/code-security/code-scanning/introduction-to-code-scanning/about-code-scanning) to our repo and would potentially detect if we try to add an insecure workflow (amongst other things).

I set it up on my fork, this shows it does protect against a user that creates a PR to add unsafe workflows like the above Github blog post:
https://github.com/lesteve/scikit-learn/pull/31

I created the workflow, clicking on the right buttons on Github. The only tweak I did is to disable code scanning for C/C++. By default C/C++ was enabled (probably because of the svm code), except that it was failing. Probably it is possible to enable it but then you need to give a custom build command. I don't think it is worth it right now.

By the way it looks like one of our js file triggers a Medium security alert (screenshot from my fork setup):
![image](https://github.com/scikit-learn/scikit-learn/assets/1680079/1b9a5713-d97f-48f7-811b-52f9bfa04e68)

lesteve committed:
MNT Add code scanning workflow

adrinjalali reviewed (commented):
Curious to see how this evolves. Thanks @lesteve

lesteve committed:
Tweak

lesteve commented:
I removed the explicit swift mentions from the automatically generated workflow and also added `*.X` branches.

I don't really know wheter this is likely to highlight many times the same issue in different `.X` branches ...

In case that happens and it's too noisy, we could as you mention start with only `1.4.X` and remember to add release branches when we create them.

betatim commented on .github/workflows/codeql.yml:
```diff
@@ -0,0 +1,73 @@
+    strategy:
+      fail-fast: false
+      matrix:
+        language: [ 'javascript-typescript', 'python' ]
```
I was trying to understand if this line means that the workflow will scan JS and Python code only?

I am somewhat interested about getting reports about our Python code, but also somewhat not excited because I assume it'll lead to false positives and "yes I know you think this is dangerous but we are adults and want to do it".

What I would be very excited about is something that vets/scans/checks the workflows and CI configs. At least I think this is where you could manage to sneak things past us or we just misconfigure things to make life easier for attackers. Do you know if something like this exists? (I think it doesn't :()

lesteve replied:
So I don't understand the reason why, but I think javascript-typescript actually also checks CI workflows ... see for example [this](https://github.blog/2023-08-09-four-tips-to-keep-your-github-actions-workflows-secure/#2-enable-code-scanning-for-workflows)

> then you need to manually modify the CodeQL workflow to add JavaScript as an additional language

I agree with you on the Python code, for example I never found LGTM very useful when it was enabled ... and I would admit that my tolerance to false positives is quite low.

The fact that there is no security report on the Python code (tested on my fork) is a good sign for me that there will not be that many false positives. We can always disable python code scanning if we agree it is on the noisy side.

Also I haven't played too much with the security report but it looks like you can tag a defect as false positive, hopefully that means that it never shows up again, see [this](https://docs.github.com/en/code-security/code-scanning/managing-your-code-scanning-configuration/about-the-tool-status-page) for more details.

glemaitre commented:
> By the way it looks like one of our js file triggers a Medium security alert

Whoops, I wrote that code. Always knew that I did not how to code in javascript :)

adrinjalali commented:
@glemaitre @betatim should we merge this then? We can always revert and disable if it gets too noisy.

lesteve commented:
Great, you can now look at security reports in https://github.com/scikit-learn/scikit-learn/security/code-scanning.
